### PR TITLE
feat: Implemented user metadata service

### PR DESCRIPTION
**Pull request by JigarJoshi**

### CODE REVIEW

File: server/services/v1/user_metadata_provider.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+	namespaceId, err := a.tenantMgr.GetNamespaceId(namespace)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to update user metadata. reason: failed to read namespace id.")
+	}
+
+	tx, err := a.txMgr.StartTx(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to update user metadata. reason: failed to create internal transaction.")
+	}
```
Comment (efirs) [inline server/services/v1/user_metadata_provider.go:125]: The code from the beginning of the function up to this point is identical in three above functions can we dedup?

File: server/services/v1/auth_provider.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+	}
+
+	// check ownership before rotating
+	currentSub, err := getCurrentSub(ctx)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to delete application: reason = %s", err.Error())
+	}
+	if client.GetClientMetadata()[createdBy] != currentSub {
+		return nil, api.Errorf(api.Code_PERMISSION_DENIED, "Failed to rotate application secret: reason = You cannot rotate secret for application that is not created by you.")
+	}
```
Comment (efirs) [inline server/services/v1/auth_provider.go:225]: The code from the beginning of the function up to this point is identical in three above functions can we dedup? (1)

File: api/server/v1/marshaler.go [hunk 3]
```diff
@@ -643,6 +643,105 @@ func (x *SubscribeResponse) MarshalJSON() ([]byte, error) {
...
+		case "metadataKey":
+			if err := jsoniter.Unmarshal(value, &x.MetadataKey); err != nil {
+				return err
+			}
+		case "value":
+			var valMp map[string]jsoniter.RawMessage
+			if err := jsoniter.Unmarshal(data, &valMp); err != nil {
+				return err
+			}
+			for k1, v1 := range valMp {
```
Comment (efirs) [inline api/server/v1/marshaler.go:662]: Why not just `k, v :=` :)

File: api/server/v1/marshaler.go [hunk 4]
```diff
@@ -643,6 +643,105 @@ func (x *SubscribeResponse) MarshalJSON() ([]byte, error) {
...
+				if k1 == "value" {
+					x.Value = &UserMetadataValue{}
```
Comment (efirs) [inline api/server/v1/marshaler.go:664]: `x.Value = &UserMetadataValue{Value: v1}`

File: api/server/v1/marshaler.go [hunk 5]
```diff
@@ -643,6 +643,105 @@ func (x *SubscribeResponse) MarshalJSON() ([]byte, error) {
 	return json.Marshal(resp)
 }
 
+func (x *InsertUserMetadataRequest) UnmarshalJSON(data []byte) error {
```
Comment (efirs) [inline api/server/v1/marshaler.go:646]: Why do we need custom unmarshalling?
Can we define `UserMetadataValue` in the proto?

File: server/services/v1/user_metadata_provider.go (hunk 1 above)
Comment (JigarJoshi) [reply server/services/v1/user_metadata_provider.go:103]: refactored.

File: server/services/v1/auth_provider.go (hunk 2 above)
Comment (JigarJoshi) [reply server/services/v1/auth_provider.go:195]: refactored.

File: api/server/v1/marshaler.go (hunk 3 above)
Comment (JigarJoshi) [reply api/server/v1/marshaler.go]: renamed variables.

File: api/server/v1/marshaler.go (hunk 4 above)
Comment (JigarJoshi) [reply api/server/v1/marshaler.go]: updated.

File: api/server/v1/marshaler.go (hunk 5 above)
Comment (JigarJoshi) [reply api/server/v1/marshaler.go:646]: value is []byte, without this it works with base64 encoded data. this is to give human readable interface to user. please let me know if there is alternate to handle this.

Comment (efirs) [reply api/server/v1/marshaler.go:646]: That makes sense.
But we still expect some structure of that object?
```
if k == "value" {
	x.Value = &UserMetadataValue{
		Value: v,
	}
}
```

File: api/server/v1/marshaler.go [hunk 6]
```diff
@@ -643,6 +643,106 @@ func (x *SubscribeResponse) MarshalJSON() ([]byte, error) {
...
+		return err
+	}
+	for key, value := range mp {
+		switch key {
+		case "metadataKey":
+			if err := jsoniter.Unmarshal(value, &x.MetadataKey); err != nil {
+				return err
+			}
+		case "value":
+			var valMp map[string]jsoniter.RawMessage
```
Comment (efirs) [inline api/server/v1/marshaler.go:658]: should this be just:
x.Value = &UserMetadataValue{
  Value: value,
}

Comment (JigarJoshi) [reply api/server/v1/marshaler.go]: synced up offline and made value to be freeform.

File: api/server/v1/marshaler.go (hunk 5 above)
Comment (JigarJoshi) [reply api/server/v1/marshaler.go:646]: synced up offline and made value to be freeform.

**Review by efirs — APPROVED** [review]

File: server/services/v1/user_metadata_provider.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	"context"
+
+	"github.com/rs/zerolog/log"
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/server/metadata"
+	"github.com/tigrisdata/tigris/server/request"
+	"github.com/tigrisdata/tigris/server/transaction"
+)
+
+type UserMetadataProvider interface {
```
Comment (himank) [inline server/services/v1/user_metadata_provider.go:13]: We need some integration tests for this provider.

File: server/services/v1/user.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/rs/zerolog/log"
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/server/config"
+	"github.com/tigrisdata/tigris/server/metadata"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"google.golang.org/grpc"
+)
+
+const (
+	userPattern = "/users/v1/*"
```
Comment (himank) [inline server/services/v1/user.go:33]: would you be changing this name pattern to be `api/v1/users`?

File: server/metadata/users.go [hunk 9]
```diff
@@ -61,10 +62,10 @@ func (u *UserSubspace) GetUserMetadata(ctx context.Context, tx transaction.Tx, n
 		return nil, err
 	}
 	var row kv.KeyValue
-	for it.Next(&row) {
+	if it.Next(&row) {
```
Comment (himank) [inline server/metadata/users.go:65]: The below `UpdateUserMetadata` still needs fixing.

File: server/services/v1/user_metadata_provider.go (hunk 7 above)
Comment (JigarJoshi) [reply server/services/v1/user_metadata_provider.go:13]: we need integration testing that covers all area behind auth/namespace yes. I will take care of integration test in a couple of week.

File: server/services/v1/user.go (hunk 8 above)
Comment (JigarJoshi) [reply server/services/v1/user.go:33]: HTTP api path refactoring will be done later once all functionality is in.

File: server/metadata/users.go (hunk 9 above)
Comment (JigarJoshi) [reply server/metadata/users.go:65]: I did not follow your suggestion here. please elaborate.

**Pull request merged**